internal/provider/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,198 @@
+package provider
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_protected_environment", func() *schema.Resource {
+	// lintignore: XR002 // TODO: Resolve this tfproviderlint issue
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage Project Access Token for your GitLab projects.",
+
+		CreateContext: resourceGitlabProjectProtectedEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectProtectedEnvironmentRead,
+		DeleteContext: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "The ID or full path of the project which the protected environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Description:  "The name of the environment.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"deploy_access_levels": {
+				Description: "Array of access levels allowed to deploy, with each described by a hash.",
+				Type:        schema.TypeList,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Description:  "Level of access required to deploy to this protected environment (developer, maintainer)",
+							Type:         schema.TypeString,
+							ForceNew:     true,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"developer", "maintainer"}, false),
```

Comment:
Can one of the arrays from the `access_level_helpers.go` be used?

Reply:
I created one for this purpose as it seems others have their own arrays as well.  Please resolve this thread if it is what you were thinking

Reply:
Okay! I'll review again if all / the majority of things have been addresses if that's okay.

Reply:
Yeah that is fine I was just thinking small reviews vs big ones because I don't want to go in the wrong direction since I don't really know Go